File: 计算机图像学/labs/cpp/lab2.cpp

#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>

//#include<opencv2/opencv.hpp>
#include<iostream>
#include<string>
#include<map>
#include<random>
#include<ctime>

using namespace std;
using namespace cv;

int mean_filter(Mat &img, int window_size = 3)
{
    if (window_size % 2 == 0)window_size += 1;

    if (img.rows < 2 * window_size || img.cols < 2 * window_size) {
        cout << "[Error]: filter window is too large!" << endl;
        exit(1);
    }

    Mat img2 = Mat::zeros(img.size(), img.type());
    int half = window_size / 2, tot = window_size * window_size;
    int m = img.rows - half, n = img.cols - half;
    long sm;

    for (int i = 0; i < img.rows; ++i)
        for (int j = 0; j < img.cols; ++j) {
            if (i < half || j < half || i >= m || j >= n) {
                img2.at<uchar>(i, j) = img.at<uchar>(i, j);
            } else {
                if (j == half) {
                    sm = 0;

                    for (int i2 = i - half; i2 <= i + half; ++i2)
                        for (int j2 = 0; j2 < window_size; ++j2)
                            sm += img.at<uchar>(i2, j2);
                } else {
                    for (int r = i - half; r <= i + half; ++r)
                        sm += img.at<uchar>(r, j + half) - img.at<uchar>(r, j - half - 1);
                }

                img2.at<uchar>(i, j) = (sm + tot - 1) / tot;
            }
        }

    imshow("filtered", img2);
    waitKey();
}

uchar find_median(vector<uchar> vec)
{
    //return the median of an array
    int mid = vec.size() / 2;
    int i = 0, j = vec.size() - 1;

    while (1) {
        int p = i;
        auto pivot = vec[p];

        while (j > i) {
            while (j > i && vec[j] > pivot)--j;

            if (j > i)
                vec[i++] = vec[j];

            while (j > i && vec[i] < pivot)++i;

            if (j > i)
                vec[j--] = vec[i];
        }

        vec[i] = pivot;

        if (i == mid)
            return pivot;
        else if (i > mid) {
            i = 0;
            j -= 1; 
        } else {
            i += 1;
            j = vec.size() - 1;
        }
    }
}
int median_filter(Mat &img, int window_size = 3)
{
    if (window_size % 2 == 0)window_size += 1;

    if (img.rows < 2 * window_size || img.cols < 2 * window_size) {
        cout << "[Error]: filter window is too large!" << endl;
        exit(1);
    }

    Mat img2 = Mat::zeros(img.size(), img.type());
    int half = window_size / 2, tot = window_size * window_size;
    int m = img.rows - half, n = img.cols - half;
    vector<uchar> queue;

    for (int i = 0; i < img.rows; ++i)
        for (int j = 0; j < img.cols; ++j) {
            if (i < half || j < half || i >= m || j >= n) {
                img2.at<uchar>(i, j) = img.at<uchar>(i, j);
            } else {
                if (j == half) {
                    queue.clear();

                    // push_back by columns
                    for (int j2 = 0; j2 < window_size; ++j2) // important
                        for (int i2 = i - half; i2 <= i + half; ++i2)
                            queue.push_back(img.at<uchar>(i2, j2));
                } else {
                    for (int k = 0, r; k < window_size; ++k) {
                        r =  i - half + k;
                        queue[k + (j + half) % window_size] = img.at<uchar>(r, j + half) ;
                    }
                }

                img2.at<uchar>(i, j) = find_median(queue);
            }
        }

    imshow("filtered", img2);
    waitKey();
}

void noise(Mat img,unsigned rate=3)
{
    default_random_engine e(time(0));
    uniform_int_distribution<unsigned> u(1, 100);// 定义一个范围为0~9的无符号整型分布类型
    uchar mn=255,mx=0; 
    for (int i = 0; i < img.rows; ++i)
        for (int j = 0; j < img.cols; ++j) {
            auto p = img.at<uchar>(i,j); 
            if(mn>p)
                mn = p;
            if(mx<p)
                mx = p;
        }

    for (int i = 0; i < img.rows; ++i)
        for (int j = 0; j < img.cols; ++j) {
            if(u(e)<=rate){
                if(u(e)<=50)
                    img.at<uchar>(i,j) = mx;
                else
                    img.at<uchar>(i,j) = mn;
            }
        }
}


int main(int argc, char** argv)
{
    Mat img = imread("images/lena.bmp", IMREAD_GRAYSCALE);

    noise(img,3); 

    namedWindow("before", 1);
    imshow("before", img);
    waitKey();
    cout << "Mean filter..." << endl;
    mean_filter(img);
    cout << "median filter..." << endl;
    median_filter(img);
    return 0;
}

File: 计算机图像学/labs/lab2.py
import sys
from random import random
from collections import Iterable

import cv2
import numpy as np
import matplotlib as mpl
from matplotlib import pyplot as plt


def noise(img, rate=0.03):
    '''3% 的椒盐噪音'''
    img2 = np.zeros(img.shape, img.dtype)
    n, m = img.shape
    mn = 255
    mx = 0
    for i in range(n):
        for j in range(m):
            if img[i][j] < mn:
                mn = img[i][j]
            if img[i][j] > mx:
                mx = img[i][j]
    for i in range(n):
        for j in range(m):
            if random() <= rate:
                if random() <= 0.5:
                    img2[i][j] = mn
                else:
                    img2[i][j] = mx
            else:
                img2[i][j] = img[i][j]
    return img2


def mean_filter(img, window_size=3):
    def mysum(obj):
        if isinstance(obj, Iterable):
            return sum((mysum(i) for i in obj),
                       0)  # intital int 0, convert uint8 to int
        return obj

    if window_size % 2 == 0:
        window_size += 1

    n, m = img.shape
    if n < 2 * window_size or m < 2 * window_size:
        raise Exception("[Error]: filter window is too large!")

    img2 = np.zeros(img.shape, img.dtype)

    half = window_size // 2
    tot = window_size * window_size

    n2 = n - half
    m2 = m - half
    sm = 0
    for i in range(n):
        for j in range(m):
            if i < half or j < half or i >= n2 or j >= m2:
                img2[i][j] = img[i][j]

            else:
                if j == half:
                    sm = mysum(img[i - half:i + half + 1, :window_size])
                else:
                    sm += mysum(img[i - half:i + half + 1, j + half]) - mysum(
                        img[i - half:i + half + 1, j - half - 1])
                img2[i][j] = (sm + tot - 1) // tot

    return img2


def median_filter(img, window_size=3):
    def find_median(arr):
        i, j = 0, len(arr) - 1
        mid = len(arr) // 2
        while 1:
            p = i
            pivot = arr[p]
            while i < j:
                while i < j and arr[j] >= pivot:
                    j -= 1
                if i < j:
                    arr[i] = arr[j]
                    i += 1
                while i < j and arr[i] <= pivot:
                    i += 1
                if i < j:
                    arr[j] = arr[i]
                    j -= 1
            arr[i] = pivot
            if i < mid:
                i += 1
                j = len(arr) - 1
            elif j > mid:
                i = 0
                j -= 1
            else:
                return pivot

    if window_size % 2 == 0:
        window_size += 1

    n, m = img.shape
    if n < 2 * window_size or m < 2 * window_size:
        raise Exception("[Error]: filter window is too large!")

    img2 = np.zeros(img.shape, img.dtype)

    half = window_size // 2

    n2 = n - half
    m2 = m - half

    beg1 = end1 = beg2 = end2 = 0
    for i in range(n):
        for j in range(m):
            if i < half or j < half or i >= n2 or j >= m2:
                img2[i][j] = img[i][j]

            else:
                if j == half:
                    # add  by columns
                    beg1 = i - half
                    end1 = i + half + 1
                    beg2 = 0
                    end2 = window_size
                else:
                    beg2 += 1
                    end2 += 1
                window = list(img[beg1:end1, beg2:end2].flat)
                img2[i][j] = find_median(window)
    return img2


if __name__ == '__main__':
    path = sys.argv[1]
    img = cv2.imread(path, cv2.IMREAD_GRAYSCALE)
    noised_img = noise(img)

    img2 = mean_filter(noised_img)
    img3 = median_filter(noised_img)

    cmap = mpl.cm.gray
    plt.figure(figsize=(8, 8))

    plt.subplot(221), plt.xticks([]), plt.yticks([])
    plt.imshow(img, cmap=cmap)
    plt.title('origin')

    plt.subplot(222), plt.xticks([]), plt.yticks([])
    plt.imshow(noised_img, cmap=cmap)
    plt.title('noised')

    plt.subplot(223), plt.xticks([]), plt.yticks([])
    plt.imshow(img2, cmap=cmap)
    plt.title('mean_filter')

    plt.subplot(224), plt.xticks([]), plt.yticks([])
    plt.imshow(img3, cmap=cmap)
    plt.title('median_filter')

    plt.show()
